File: Services/Concrete/Excel/Sheets/Single/GeneralSheet.cs
using System;
using System.Collections.Generic;
using Core.Models;

namespace Services.Concrete.Excel.Sheets.Single
{
    internal class GeneralSheet: SingleDemoSheet
    {
        protected override string GetName()
        {
            return "General";
        }

        protected override string[] GetColumnNames()
        {
            return new[]
            {
                "Filename",
                "ID",
                "Date",
                "Type",
                "Source",
                "Map",
                "Hostname",
                "Client",
                "Server Tickrate",
                "Framerate",
                "Duration",
                "Ticks",
                "Name team 1",
                "Name team 2",
                "Score team 1",
                "Score team 2",
                "Score 1st half team 1",
                "Score 1st half team 2",
                "Score 2nd half team 1",
                "Score 2nd half team 2",
                "Winner",
                "Kills",
                "Assists",
                "5K",
                "4K",
                "3K",
                "2K",
                "1K",
                "Trade kill",
                "Average Damage Per Round",
                "Total Damage Health",
                "Total Damage Armor",
                "Average KAST",
                "Clutch",
                "Bomb Defused",
                "Bomb Exploded",
                "Bomb Planted",
                "Flashbang",
                "Smoke",
                "HE",
                "Decoy",
                "Molotov",
                "Incendiary",
                "Shots",
                "Hits",
                "Round",
                "Comment",
                "Cheater",
            };
        }

        public GeneralSheet(Workbook workbook, Demo demo): base(workbook, demo)
        {
        }

        public override void Generate()
        {
            var cells = new List<object>
            {
                Demo.Name,
                Demo.Id,
                Demo.DateAsString,
                Demo.Type.AsString(),
                Demo.Source.Name,
                Demo.MapName,
                Demo.Hostname,
                Demo.ClientName,
                Demo.ServerTickrate,
                Demo.Tickrate,
                Math.Round(Demo.Duration, 2),
                Demo.Ticks,
                Demo.TeamCT.Name,
                Demo.TeamT.Name,
                Demo.ScoreTeamCt,
                Demo.ScoreTeamT,
                Demo.ScoreFirstHalfTeamCt,
                Demo.ScoreFirstHalfTeamT,
                Demo.ScoreSecondHalfTeamCt,
                Demo.ScoreSecondHalfTeamT,
                Demo.Winner != null ? Demo.Winner.Name : string.Empty,
                Demo.KillCount,
                Demo.AssistCount,
                Demo.FiveKillCount,
                Demo.FourKillCount,
                Demo.ThreeKillCount,
                Demo.TwoKillCount,
                Demo.OneKillCount,
                Demo.TradeKillCount,
                Demo.AverageHealthDamage,
                Demo.DamageHealthCount,
                Demo.DamageArmorCount,
                Math.Round(Demo.AverageKast, 2),
                Demo.ClutchCount,
                Demo.BombDefusedCount,
                Demo.BombExplodedCount,
                Demo.BombPlantedCount,
                Demo.FlashbangThrownCount,
                Demo.SmokeThrownCount,
                Demo.HeThrownCount,
                Demo.DecoyThrownCount,
                Demo.MolotovThrownCount,
                Demo.IncendiaryThrownCount,
                Demo.WeaponFiredCount,
                Demo.HitCount,
                Demo.Rounds.Count,
                Demo.Comment,
                Demo.CheaterCount,
            };
            WriteRow(cells);
        }
    }
}

File: src/node/xlsx/match-export/general-sheet.ts
import type { Match } from 'csdm/common/types/match';
import type { Column } from '../column';
import { SingleMatchExportSheet } from './single-match-export-sheet';

export class GeneralSheet extends SingleMatchExportSheet<Match> {
  protected getName() {
    return 'General';
  }

  protected getColumns(): Column<Match>[] {
    return [
      {
        name: 'checksum',
        cellFormatter: (row) => row.checksum,
      },
      {
        name: 'name',
        cellFormatter: (row) => row.name,
      },
      {
        name: 'demo_path',
        cellFormatter: (row) => row.demoFilePath,
      },
      {
        name: 'date',
        cellFormatter: (row) => new Date(row.date),
      },
      {
        name: 'source',
        cellFormatter: (row) => row.source,
      },
      {
        name: 'map',
        cellFormatter: (row) => row.mapName,
      },
      {
        name: 'server_name',
        cellFormatter: (row) => row.serverName,
      },
      {
        name: 'clientname',
        cellFormatter: (row) => row.clientName,
      },
      {
        name: 'duration',
        cellFormatter: (row) => row.duration,
      },
      {
        name: 'tickrate',
        cellFormatter: (row) => row.tickrate,
      },
      {
        name: 'tick_count',
        cellFormatter: (row) => row.tickCount,
      },
      {
        name: 'frame_rate',
        cellFormatter: (row) => row.frameRate,
      },
      {
        name: 'name_team_a',
        cellFormatter: (row) => row.teamA.name,
      },
      {
        name: 'name_team_b',
        cellFormatter: (row) => row.teamB.name,
      },
      {
        name: 'score_team_a',
        cellFormatter: (row) => row.teamA.score,
      },
      {
        name: 'score_team_b',
        cellFormatter: (row) => row.teamB.score,
      },
      {
        name: 'kill_count',
        cellFormatter: (row) => row.killCount,
      },
      {
        name: 'assist_count',
        cellFormatter: (row) => row.assistCount,
      },
      {
        name: 'death_count',
        cellFormatter: (row) => row.deathCount,
      },
      {
        name: 'bomb_planted_count',

        cellFormatter: (row) => row.bombsPlanted.length,
      },
      {
        name: 'bomb_defused_count',

        cellFormatter: (row) => row.bombsDefused.length,
      },
      {
        name: 'clutch_count',
        cellFormatter: (row) => row.clutches.length,
      },
    ];
  }

  public generate() {
    this.writeRow(this.match);
  }
}
